crypto/x509/x509_vfy.c

```diff
@@ -3440,6 +3444,29 @@ static int check_key_level(X509_STORE_CTX *ctx, X509 *cert)
     return EVP_PKEY_security_bits(pkey) >= minbits_table[level - 1];
 }
 
+/*
+ * Check whether the public key of ``cert`` does not use explicit params
+ * for an elliptic curve.
+ *
+ * Returns 1 on success, 0 otherwise.
+ */
+static int check_curve(X509 *cert)
+{
+#ifndef OPENSSL_NO_EC
+    EVP_PKEY *pkey = X509_get0_pubkey(cert);
+
+    /* Unsupported or malformed keys will fail other checks, ignore them */
+    if (pkey == NULL)
+        return 1;
+
+    if (EVP_PKEY_id(pkey) == EVP_PKEY_EC
+        && EC_KEY_decoded_from_explicit_params(EVP_PKEY_get0_EC_KEY(pkey)))
```

Comment:
A trivial refractor would be needed if this can also return -1 for malformed arguments.